File: packages/go/exports.go
package cnos

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strings"
	"time"
)

type ToEnvOptions struct {
	IncludeSecrets bool
}

type ToPublicEnvOptions struct {
	Framework string
	Prefix    string
}

func (runtime *Runtime) ReadOr(key string, fallback any) (any, error) {
	value, ok, err := runtime.Read(key)
	if err != nil {
		return nil, err
	}
	if !ok {
		return fallback, nil
	}
	return value, nil
}

func (runtime *Runtime) ToObject() (map[string]any, error) {
	return runtime.toNamespaceObject("")
}

func (runtime *Runtime) ToNamespace(namespace string) (map[string]any, error) {
	return runtime.toNamespaceObject(strings.TrimSpace(namespace))
}

func (runtime *Runtime) ToEnv(options ...ToEnvOptions) (map[string]string, error) {
	config := ToEnvOptions{}
	if len(options) > 0 {
		config = options[0]
	}

	output := map[string]string{}
	keys := make([]string, 0, len(runtime.manifest.EnvMapping.Explicit))
	for envVar := range runtime.manifest.EnvMapping.Explicit {
		keys = append(keys, envVar)
	}
	sort.Strings(keys)

	for _, envVar := range keys {
		logicalKey := runtime.manifest.EnvMapping.Explicit[envVar]
		entry := runtime.entries[logicalKey]
		if entry == nil {
			continue
		}

		definition := runtime.namespaceDefinition(entry.namespace)
		if definition.Kind != "data" {
			continue
		}
		if entry.namespace == "secret" {
			if !config.IncludeSecrets {
				continue
			}
		} else if !definition.Shareable || definition.Sensitive {
			continue
		}

		value, ok, err := runtime.Read(logicalKey)
		if err != nil {
			return nil, err
		}
		if !ok || value == nil {
			continue
		}
		output[envVar] = normalizeEnvValue(value)
	}

	return output, nil
}

func (runtime *Runtime) ToPublicEnv(options ...ToPublicEnvOptions) (map[string]string, error) {
	config := ToPublicEnvOptions{}
	if len(options) > 0 {
		config = options[0]
	}

	prefix, err := runtime.resolvePublicPrefix(config)
	if err != nil {
		return nil, err
	}

	output := map[string]string{}
	keys := make([]string, 0)
	for key, entry := range runtime.entries {
		if entry.namespace == "public" {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)

	for _, key := range keys {
		sourceKey := runtime.resolveProjectedSourceKey(key)
		if source := runtime.entries[sourceKey]; source != nil && source.formula != nil && source.formula.runtimeDependent {
			value, ok, err := runtime.Read(key)
			if err != nil {
				return nil, err
			}
			if !ok || value == nil {
				return nil, fmt.Errorf("cnos: cannot build public output for %s because it depends on runtime-only values", key)
			}
		}

		value, ok, err := runtime.Read(key)
		if err != nil {
			return nil, err
		}
		if !ok || value == nil {
			continue
		}

		baseEnvVar := fallbackPublicEnvVar(strings.TrimPrefix(key, "public."))
		envVar := baseEnvVar
		if prefix != "" && !strings.HasPrefix(baseEnvVar, prefix) {
			envVar = prefix + baseEnvVar
		}
		output[envVar] = normalizeEnvValue(value)
	}

	return output, nil
}

func (runtime *Runtime) ToServerProjection() (ServerProjection, error) {
	if runtime.graphBootstrapped {
		return ServerProjection{}, fmt.Errorf("cnos: runtime graph bootstrap payload does not support server projection export")
	}
	if runtime.projection.Version == 1 && runtime.projection.ResolvedAt != "" && runtime.projection.ConfigHash != "" {
		return runtime.projection, nil
	}

	values := map[string]any{}
	derived := map[string]DerivedFormula{}
	secretRefs := map[string]SecretReference{}
	publicKeys := make([]string, 0)
	namespaces := map[string]struct{}{}
	runtimeNamespaces := map[string]struct{}{}

	keys := make([]string, 0, len(runtime.entries))
	for key := range runtime.entries {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	for _, key := range keys {
		entry := runtime.entries[key]
		if entry == nil || entry.namespace == "meta" || entry.namespace == "public" {
			continue
		}

		if entry.secretRef != nil {
			ref := *entry.secretRef
			if ref.Provider == "" {
				ref.Provider = runtime.secretVaultDefinition(ref).Provider
			}
			if ref.EnvVar == "" {
				ref.EnvVar = runtime.logicalRefToMappedEnvVar(ref.Vault, ref.Ref)
			}
			secretRefs[strings.TrimPrefix(key, "secret.")] = ref
			continue
		}

		definition := runtime.namespaceDefinition(entry.namespace)
		if definition.Kind != "data" || definition.Sensitive {
			continue
		}
		if runtime.sources[key] == "process-env" {
			continue
		}

		projectedKey := key
		if entry.namespace == "value" {
			projectedKey = strings.TrimPrefix(key, "value.")
		} else {
			namespaces[entry.namespace] = struct{}{}
		}

		if entry.formula != nil {
			if entry.formula.runtimeDependent {
				derived[projectedKey] = DerivedFormula{
					Expr:        entry.formula.raw,
					Deps:        append([]string(nil), entry.formula.deps...),
					RuntimeRefs: append([]string(nil), entry.formula.runtimeRefs...),
				}
				for _, ref := range entry.formula.runtimeRefs {
					namespace := namespaceForKey(ref)
					if namespace != "" {
						runtimeNamespaces[namespace] = struct{}{}
					}
				}
				continue
			}
		}

		value, ok, err := runtime.Read(key)
		if err != nil {
			return ServerProjection{}, err
		}
		if !ok {
			continue
		}
		values[projectedKey] = value
	}

	for _, key := range keys {
		entry := runtime.entries[key]
		if entry != nil && entry.namespace == "public" {
			publicKeys = append(publicKeys, strings.TrimPrefix(key, "public."))
		}
	}

	namespaceList := sortedStringSet(namespaces)
	runtimeNamespaceList := sortedStringSet(runtimeNamespaces)
	projection := ServerProjection{
		Version:           1,
		Workspace:         runtime.profileWorkspace("workspace"),
		Profile:           runtime.profileWorkspace("profile"),
		ResolvedAt:        firstNonEmpty(runtime.projection.ResolvedAt),
		ConfigHash:        configHash(values),
		Values:            stableSortAnyMap(values),
		Derived:           stableSortFormulaMap(derived),
		SecretRefs:        stableSortSecretRefMap(secretRefs),
		Vaults:            stableSortVaultMap(projectReferencedVaults(secretRefs, runtime.manifest.Vaults)),
		PublicKeys:        publicKeys,
		RuntimeNamespaces: runtimeNamespaceList,
		Meta: ProjectionMeta{
			Workspace:   runtime.profileWorkspace("workspace"),
			Profile:     runtime.profileWorkspace("profile"),
			CnosVersion: firstNonEmpty(runtime.projection.Meta.CnosVersion, "authoring-runtime"),
			Namespaces:  namespaceList,
		},
	}
	if projection.ResolvedAt == "" {
		projection.ResolvedAt = time.Now().UTC().Format(time.RFC3339)
	}
	runtime.projection = projection
	return projection, nil
}

func (runtime *Runtime) Format(message string) (string, error) {
	return formatRuntimeMessage(runtime, message)
}

func (runtime *Runtime) Log(message string) (string, error) {
	formatted, err := formatRuntimeMessage(runtime, message)
	if err != nil {
		return "", err
	}
	fmt.Println(formatted)
	return formatted, nil
}

func (runtime *Runtime) toNamespaceObject(namespace string) (map[string]any, error) {
	output := map[string]any{}
	keys := make([]string, 0, len(runtime.entries))
	for key := range runtime.entries {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	for _, key := range keys {
		entry := runtime.entries[key]
		if entry == nil {
			continue
		}
		if namespace != "" && entry.namespace != namespace {
			continue
		}

		value, ok, err := runtime.Read(key)
		if err != nil {
			return nil, err
		}
		if !ok {
			continue
		}

		targetPath := key
		if namespace != "" {
			targetPath = strings.TrimPrefix(key, namespace+".")
		}
		setNestedValue(output, strings.Split(targetPath, "."), value)
	}

	return output, nil
}

func (runtime *Runtime) namespaceDefinition(namespace string) namespaceDefinition {
	if definition, ok := runtime.manifest.Namespaces[namespace]; ok {
		return definition
	}
	if definition, ok := defaultNamespaceDefs[namespace]; ok {
		return definition
	}
	return namespaceDefinition{Kind: "data"}
}

func (runtime *Runtime) resolvePublicPrefix(options ToPublicEnvOptions) (string, error) {
	if options.Prefix != "" {
		return options.Prefix, nil
	}
	if options.Framework == "" {
		return "", nil
	}
	prefix, ok := runtime.manifest.Frameworks[options.Framework]
	if !ok {
		return "", fmt.Errorf("cnos: unknown public framework prefix: %s", options.Framework)
	}
	return prefix, nil
}

func (runtime *Runtime) resolveProjectedSourceKey(key string) string {
	if entry := runtime.entries[key]; entry != nil {
		if entry.aliasTo != "" {
			return entry.aliasTo
		}
		if entry.promotedFrom != "" {
			return entry.promotedFrom
		}
	}
	if strings.HasPrefix(key, "public.") {
		fallback := "value." + strings.TrimPrefix(key, "public.")
		if runtime.entries[fallback] != nil {
			return fallback
		}
	}
	return key
}

func (runtime *Runtime) logicalRefToMappedEnvVar(vaultID, ref string) string {
	mapping := runtime.manifest.Vaults[vaultID].Mapping
	for envVar, logicalRef := range mapping {
		if logicalRef == ref {
			return envVar
		}
	}
	return ""
}

func (runtime *Runtime) profileWorkspace(kind string) string {
	switch kind {
	case "workspace":
		if value, ok, err := runtime.Meta("workspace"); err == nil && ok {
			if text, ok := value.(string); ok {
				return text
			}
		}
	case "profile":
		if value, ok, err := runtime.Meta("profile"); err == nil && ok {
			if text, ok := value.(string); ok {
				return text
			}
		}
	case "resolvedAt":
		return runtime.projection.ResolvedAt
	}
	return ""
}

func bootstrappedManifestFromProjection(projection ServerProjection) authoringManifest {
	namespaces := map[string]namespaceDefinition{}
	for key, value := range defaultNamespaceDefs {
		namespaces[key] = value
	}
	for _, namespace := range projection.Meta.Namespaces {
		if _, ok := namespaces[namespace]; !ok {
			namespaces[namespace] = namespaceDefinition{Kind: "data"}
		}
	}

	runtimeNamespaces := map[string]runtimeNamespaceDefinition{
		"process": {
			Description: "Live process runtime values.",
			ServerOnly:  true,
			BuiltIn:     true,
		},
	}
	for _, namespace := range projection.RuntimeNamespaces {
		if namespace == "process" {
			continue
		}
		runtimeNamespaces[namespace] = runtimeNamespaceDefinition{ServerOnly: true}
	}

	frameworks := map[string]string{}
	for key, value := range defaultFrameworks {
		frameworks[key] = value
	}

	return authoringManifest{
		ProjectName:       "bootstrapped",
		EnvMapping:        envMappingConfig{Explicit: map[string]string{}},
		Frameworks:        frameworks,
		Namespaces:        namespaces,
		RuntimeNamespaces: runtimeNamespaces,
		Vaults:            stableSortVaultMap(projection.Vaults),
	}
}

func formatRuntimeMessage(runtime *Runtime, message string) (string, error) {
	var err error
	formatted := templatePattern.ReplaceAllStringFunc(message, func(token string) string {
		if err != nil {
			return token
		}
		match := templatePattern.FindStringSubmatch(token)
		if len(match) < 2 {
			return token
		}
		key := strings.TrimSpace(match[1])
		if key == "" {
			return token
		}
		value, ok, readErr := runtime.Read(key)
		if readErr != nil {
			err = readErr
			return token
		}
		if !ok {
			return token
		}
		return stringifyLogValue(value)
	})
	if err != nil {
		return "", err
	}
	return formatted, nil
}

func setNestedValue(target map[string]any, pathSegments []string, value any) {
	if len(pathSegments) == 0 || pathSegments[0] == "" {
		return
	}
	head := pathSegments[0]
	if len(pathSegments) == 1 {
		target[head] = value
		return
	}

	current, ok := target[head].(map[string]any)
	if !ok {
		current = map[string]any{}
		target[head] = current
	}
	setNestedValue(current, pathSegments[1:], value)
}

func normalizeEnvValue(value any) string {
	return jsStringifyValue(value)
}

func fallbackPublicEnvVar(valuePath string) string {
	var builder strings.Builder
	lastUnderscore := false
	for index, char := range valuePath {
		switch {
		case char >= 'a' && char <= 'z':
			if index > 0 && lastWasLowerAlphaNum(valuePath, index-1) && isUpperAhead(valuePath, index) {
				if !lastUnderscore {
					builder.WriteByte('_')
				}
			}
			builder.WriteRune(char - 32)
			lastUnderscore = false
		case char >= 'A' && char <= 'Z':
			if index > 0 && lastWasLowerAlphaNum(valuePath, index-1) && !lastUnderscore {
				builder.WriteByte('_')
			}
			builder.WriteRune(char)
			lastUnderscore = false
		case char >= '0' && char <= '9':
			builder.WriteRune(char)
			lastUnderscore = false
		default:
			if !lastUnderscore {
				builder.WriteByte('_')
				lastUnderscore = true
			}
		}
	}
	return strings.Trim(builder.String(), "_")
}

func lastWasLowerAlphaNum(value string, index int) bool {
	if index < 0 || index >= len(value) {
		return false
	}
	char := value[index]
	return (char >= 'a' && char <= 'z') || (char >= '0' && char <= '9')
}

func isUpperAhead(value string, index int) bool {
	if index+1 >= len(value) {
		return false
	}
	next := value[index+1]
	return next >= 'A' && next <= 'Z'
}

func configHash(values map[string]any) string {
	sum := sha256.Sum256([]byte(stableJSONString(values)))
	return hex.EncodeToString(sum[:])
}

func stableJSONString(value map[string]any) string {
	encoded, err := json.Marshal(stableSortAnyMap(value))
	if err != nil {
		return "{}"
	}
	return string(encoded)
}

func stableSortAnyMap(value map[string]any) map[string]any {
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	sorted := map[string]any{}
	for _, key := range keys {
		sorted[key] = stableSortAnyValue(value[key])
	}
	return sorted
}

func stableSortAnyValue(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		return stableSortAnyMap(typed)
	case []any:
		items := make([]any, len(typed))
		for index, item := range typed {
			items[index] = stableSortAnyValue(item)
		}
		return items
	default:
		return value
	}
}

func stableSortFormulaMap(value map[string]DerivedFormula) map[string]DerivedFormula {
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	sorted := map[string]DerivedFormula{}
	for _, key := range keys {
		sorted[key] = value[key]
	}
	return sorted
}

func stableSortSecretRefMap(value map[string]SecretReference) map[string]SecretReference {
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	sorted := map[string]SecretReference{}
	for _, key := range keys {
		sorted[key] = value[key]
	}
	return sorted
}

func stableSortStringMap(value map[string]string) map[string]string {
	if len(value) == 0 {
		return nil
	}
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	sorted := map[string]string{}
	for _, key := range keys {
		sorted[key] = value[key]
	}
	return sorted
}

func projectReferencedVaults(secretRefs map[string]SecretReference, vaults map[string]vaultDefinition) map[string]vaultDefinition {
	projected := map[string]vaultDefinition{}
	for _, ref := range secretRefs {
		if definition, ok := vaults[ref.Vault]; ok {
			projected[ref.Vault] = projectVaultDefinition(definition)
		}
	}
	return projected
}

func projectVaultDefinition(definition vaultDefinition) vaultDefinition {
	projected := vaultDefinition{
		Provider: definition.Provider,
		Mapping:  stableSortStringMap(definition.Mapping),
	}

	projected.Auth = projectVaultAuth(definition.Auth)
	for _, fallback := range definition.Fallback {
		projected.Fallback = append(projected.Fallback, projectVaultDefinition(fallback))
	}
	return projected
}

func projectVaultAuth(auth vaultAuthFile) vaultAuthFile {
	return vaultAuthFile{
		Method:     auth.Method,
		Passphrase: cloneVaultAuthSource(auth.Passphrase),
		Token:      cloneVaultAuthSource(auth.Token),
		Config:     sanitizeProjectedConfig(auth.Config),
	}
}

func cloneVaultAuthSource(source *vaultAuthSourceFile) *vaultAuthSourceFile {
	if source == nil {
		return nil
	}
	return &vaultAuthSourceFile{From: append([]string(nil), source.From...)}
}

var safeProjectedConfigKeys = map[string]struct{}{
	"address":             {},
	"audience":            {},
	"clientid":            {},
	"endpoint":            {},
	"mount":               {},
	"namespace":           {},
	"path":                {},
	"projectid":           {},
	"region":              {},
	"scope":               {},
	"scopes":              {},
	"serviceaccountemail": {},
	"tenant":              {},
	"tenantid":            {},
	"url":                 {},
	"version":             {},
	"vaulturl":            {},
}

func sanitizeProjectedConfig(config map[string]any) map[string]any {
	sanitized, ok := sanitizeProjectedConfigValue(config).(map[string]any)
	if !ok || len(sanitized) == 0 {
		return nil
	}
	return sanitized
}

func sanitizeProjectedConfigValue(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		output := map[string]any{}
		for key, item := range typed {
			sanitized := sanitizeProjectedConfigValue(item)
			if nested, ok := sanitized.(map[string]any); ok {
				if len(nested) > 0 {
					output[key] = nested
				}
				continue
			}
			if _, ok := safeProjectedConfigKeys[normalizeProjectedConfigKey(key)]; ok {
				output[key] = sanitized
			}
		}
		return stableSortAnyMap(output)
	case []any:
		items := make([]any, len(typed))
		for index, item := range typed {
			items[index] = sanitizeProjectedConfigValue(item)
		}
		return items
	default:
		return value
	}
}

func normalizeProjectedConfigKey(key string) string {
	var builder strings.Builder
	for _, char := range key {
		if (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || (char >= '0' && char <= '9') {
			builder.WriteRune(char)
		}
	}
	return strings.ToLower(builder.String())
}

func stableSortVaultMap(value map[string]vaultDefinition) map[string]vaultDefinition {
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	sorted := map[string]vaultDefinition{}
	for _, key := range keys {
		sorted[key] = value[key]
	}
	return sorted
}

func sortedStringSet(values map[string]struct{}) []string {
	items := make([]string, 0, len(values))
	for value := range values {
		if value != "" {
			items = append(items, value)
		}
	}
	sort.Strings(items)
	return items
}

func stringifyLogValue(value any) string {
	return jsLogStringifyValue(value)
}

var templatePattern = regexp.MustCompile(`\$\{([^}]+)\}`)

File: packages/python/cnos/src/cnos/exports.py
"""Export helpers — to_object, to_namespace, to_env, to_public_env, to_server_projection,
format_message, log_message, config_hash. Mirrors Go's exports.go.
These are implemented as free functions that accept a CnosRuntime instance.
"""
from __future__ import annotations

import hashlib
import json
import re
import time
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Set, Tuple

from cnos.errors import CnosError
from cnos.jscompat import js_log_stringify_value, js_stringify_value
from cnos.projection import ServerProjection
from cnos.types import (
    DerivedFormula,
    SecretReference,
    ToEnvOptions,
    ToPublicEnvOptions,
    VaultDefinition,
)

if TYPE_CHECKING:
    from cnos.runtime import CnosRuntime

_TEMPLATE_PATTERN = re.compile(r"\$\{([^}]+)\}")

_SAFE_PROJECTED_CONFIG_KEYS: Set[str] = {
    "address", "audience", "clientid", "endpoint", "mount", "namespace",
    "path", "projectid", "region", "scope", "scopes", "serviceaccountemail",
    "tenant", "tenantid", "url", "version", "vaulturl",
}


# ---------------------------------------------------------------------------
# to_object / to_namespace
# ---------------------------------------------------------------------------

def to_object(runtime: "CnosRuntime") -> Dict[str, Any]:
    return _to_namespace_object(runtime, "")


def to_namespace(runtime: "CnosRuntime", namespace: str) -> Dict[str, Any]:
    return _to_namespace_object(runtime, namespace.strip())


def _to_namespace_object(runtime: "CnosRuntime", namespace: str) -> Dict[str, Any]:
    output: Dict[str, Any] = {}
    keys = sorted(runtime._entries.keys())
    for key in keys:
        entry = runtime._entries[key]
        if entry is None:
            continue
        if namespace and entry.namespace != namespace:
            continue
        value, ok = runtime.read(key)
        if not ok:
            continue
        target_path = key if not namespace else key[len(namespace) + 1:]
        _set_nested_value(output, target_path.split("."), value)
    return output


def _set_nested_value(target: Dict[str, Any], path: List[str], value: Any) -> None:
    if not path or not path[0]:
        return
    head = path[0]
    if len(path) == 1:
        target[head] = value
        return
    current = target.get(head)
    if not isinstance(current, dict):
        current = {}
        target[head] = current
    _set_nested_value(current, path[1:], value)


# ---------------------------------------------------------------------------
# to_env
# ---------------------------------------------------------------------------

def to_env(runtime: "CnosRuntime", options: Optional[ToEnvOptions] = None) -> Dict[str, str]:
    config = options or ToEnvOptions()
    output: Dict[str, str] = {}
    explicit: Dict[str, str] = runtime._manifest.get("env_mapping", {}).get("explicit", {})
    for env_var in sorted(explicit.keys()):
        logical_key = explicit[env_var]
        entry = runtime._entries.get(logical_key)
        if entry is None:
            continue
        definition = runtime._namespace_definition(entry.namespace)
        if definition.get("kind") != "data":
            continue
        if entry.namespace == "secret":
            if not config.include_secrets:
                continue
        elif not definition.get("shareable") or definition.get("sensitive"):
            continue
        value, ok = runtime.read(logical_key)
        if not ok or value is None:
            continue
        output[env_var] = _normalize_env_value(value)
    return output


# ---------------------------------------------------------------------------
# to_public_env
# ---------------------------------------------------------------------------

def to_public_env(
    runtime: "CnosRuntime", options: Optional[ToPublicEnvOptions] = None
) -> Dict[str, str]:
    config = options or ToPublicEnvOptions()
    prefix = _resolve_public_prefix(runtime, config)
    output: Dict[str, str] = {}
    keys = sorted(
        k for k, e in runtime._entries.items() if e is not None and e.namespace == "public"
    )
    for key in keys:
        source_key = runtime._resolve_projected_source_key(key)
        source = runtime._entries.get(source_key)
        if source is not None and source.formula is not None and source.formula.runtime_dependent:
            value, ok = runtime.read(key)
            if not ok or value is None:
                raise CnosError(
                    f"cnos: cannot build public output for {key} because it depends on runtime-only values"
                )
        value, ok = runtime.read(key)
        if not ok or value is None:
            continue
        base_env_var = _fallback_public_env_var(key[len("public."):])
        env_var = base_env_var
        if prefix and not base_env_var.startswith(prefix):
            env_var = prefix + base_env_var
        output[env_var] = _normalize_env_value(value)
    return output


def _resolve_public_prefix(runtime: "CnosRuntime", options: ToPublicEnvOptions) -> str:
    if options.prefix:
        return options.prefix
    if not options.framework:
        return ""
    frameworks = runtime._manifest.get("frameworks", {})
    if options.framework not in frameworks:
        raise CnosError(f"cnos: unknown public framework prefix: {options.framework}")
    return frameworks[options.framework]


def _fallback_public_env_var(value_path: str) -> str:
    """Convert a dot-path to UPPER_SNAKE_CASE env var name. Mirrors Go's fallbackPublicEnvVar."""
    parts: List[str] = []
    last_underscore = False
    for i, ch in enumerate(value_path):
        if "a" <= ch <= "z":
            # Check if next char is uppercase (camelCase transition)
            if i > 0 and _last_was_lower_alpha_num(value_path, i - 1) and _is_upper_ahead(value_path, i):
                if not last_underscore:
                    parts.append("_")
            parts.append(ch.upper())
            last_underscore = False
        elif "A" <= ch <= "Z":
            if i > 0 and _last_was_lower_alpha_num(value_path, i - 1) and not last_underscore:
                parts.append("_")
            parts.append(ch)
            last_underscore = False
        elif "0" <= ch <= "9":
            parts.append(ch)
            last_underscore = False
        else:
            if not last_underscore:
                parts.append("_")
                last_underscore = True
    result = "".join(parts).strip("_")
    return result


def _last_was_lower_alpha_num(value: str, index: int) -> bool:
    if index < 0 or index >= len(value):
        return False
    ch = value[index]
    return ("a" <= ch <= "z") or ("0" <= ch <= "9")


def _is_upper_ahead(value: str, index: int) -> bool:
    if index + 1 >= len(value):
        return False
    return "A" <= value[index + 1] <= "Z"


# ---------------------------------------------------------------------------
# to_server_projection
# ---------------------------------------------------------------------------

def to_server_projection(runtime: "CnosRuntime") -> ServerProjection:
    if runtime._graph_bootstrapped:
        raise CnosError(
            "cnos: runtime graph bootstrap payload does not support server projection export"
        )
    proj = runtime._projection
    if proj.version == 1 and proj.resolved_at and proj.config_hash:
        return proj

    values: Dict[str, Any] = {}
    derived: Dict[str, DerivedFormula] = {}
    secret_refs: Dict[str, SecretReference] = {}
    public_keys: List[str] = []
    namespaces: Set[str] = set()
    rt_namespaces: Set[str] = set()

    keys = sorted(runtime._entries.keys())
    for key in keys:
        entry = runtime._entries.get(key)
        if entry is None or entry.namespace in ("meta", "public"):
            continue
        if entry.secret_ref is not None:
            ref = SecretReference(
                ref=entry.secret_ref.ref,
                provider=entry.secret_ref.provider or runtime._secret_vault_definition(entry.secret_ref).provider,
                vault=entry.secret_ref.vault,
                env_var=entry.secret_ref.env_var or runtime._logical_ref_to_mapped_env_var(entry.secret_ref.vault, entry.secret_ref.ref),
            )
            secret_refs[key[len("secret."):]] = ref
            continue
        definition = runtime._namespace_definition(entry.namespace)
        if definition.get("kind") != "data" or definition.get("sensitive"):
            continue
        if runtime._sources.get(key) == "process-env":
            continue
        projected_key = key
        if entry.namespace == "value":
            projected_key = key[len("value."):]
        else:
            namespaces.add(entry.namespace)

        if entry.formula is not None:
            if entry.formula.runtime_dependent:
                derived[projected_key] = DerivedFormula(
                    expr=entry.formula.raw,
                    deps=list(entry.formula.deps),
                    runtime_refs=list(entry.formula.runtime_refs),
                )
                for ref in entry.formula.runtime_refs:
                    ns = ref.split(".")[0] if "." in ref else ""
                    if ns:
                        rt_namespaces.add(ns)
                continue

        value, ok = runtime.read(key)
        if not ok:
            continue
        values[projected_key] = value

    for key in keys:
        entry = runtime._entries.get(key)
        if entry is not None and entry.namespace == "public":
            public_keys.append(key[len("public."):])

    namespace_list = sorted(n for n in namespaces if n)
    rt_namespace_list = sorted(n for n in rt_namespaces if n)

    # Collect referenced vaults
    referenced_vaults: Dict[str, VaultDefinition] = {}
    for ref in secret_refs.values():
        if ref.vault and ref.vault in runtime._vaults:
            referenced_vaults[ref.vault] = runtime._vaults[ref.vault]

    resolved_at = proj.resolved_at
    if not resolved_at:
        resolved_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())

    new_proj = ServerProjection(
        version=1,
        workspace=runtime._profile_workspace("workspace") or proj.workspace,
        profile=runtime._profile_workspace("profile") or proj.profile,
        resolved_at=resolved_at,
        config_hash=config_hash(values),
        values=_stable_sort_any_map(values),
        derived={k: derived[k] for k in sorted(derived)},
        secret_refs={k: secret_refs[k] for k in sorted(secret_refs)},
        vaults={k: referenced_vaults[k] for k in sorted(referenced_vaults)},
        public_keys=public_keys,
        runtime_namespaces=rt_namespace_list,
        meta=runtime._projection.meta.__class__(
            workspace=runtime._profile_workspace("workspace") or proj.workspace,
            profile=runtime._profile_workspace("profile") or proj.profile,
            cnos_version=_first_non_empty(proj.meta.cnos_version, "authoring-runtime"),
            namespaces=namespace_list,
        ),
    )
    runtime._projection = new_proj
    return new_proj


def _first_non_empty(*values: str) -> str:
    for v in values:
        if v and v.strip():
            return v.strip()
    return ""


# ---------------------------------------------------------------------------
# format_message / log_message
# ---------------------------------------------------------------------------

def format_message(runtime: "CnosRuntime", message: str) -> str:
    error: Optional[str] = None

    def replacer(match: re.Match) -> str:
        nonlocal error
        if error:
            return match.group(0)
        key = match.group(1).strip()
        if not key:
            return match.group(0)
        value, ok = runtime.read(key)
        if not ok:
            return match.group(0)
        return js_log_stringify_value(value)

    result = _TEMPLATE_PATTERN.sub(replacer, message)
    return result


def log_message(runtime: "CnosRuntime", message: str) -> str:
    formatted = format_message(runtime, message)
    print(formatted)
    return formatted


# ---------------------------------------------------------------------------
# config_hash
# ---------------------------------------------------------------------------

def config_hash(values: Dict[str, Any]) -> str:
    stable = _stable_json_string(values)
    return hashlib.sha256(stable.encode("utf-8")).hexdigest()


def _stable_json_string(values: Dict[str, Any]) -> str:
    try:
        return json.dumps(_stable_sort_any_map(values), separators=(",", ":"), sort_keys=True)
    except (TypeError, ValueError):
        return "{}"


def _stable_sort_any_map(value: Dict[str, Any]) -> Dict[str, Any]:
    sorted_map: Dict[str, Any] = {}
    for k in sorted(value.keys()):
        sorted_map[k] = _stable_sort_any_value(value[k])
    return sorted_map


def _stable_sort_any_value(value: Any) -> Any:
    if isinstance(value, dict):
        return _stable_sort_any_map(value)
    if isinstance(value, list):
        return [_stable_sort_any_value(item) for item in value]
    return value


def _normalize_env_value(value: Any) -> str:
    return js_stringify_value(value)
